rust/private/stamp.bzl

```diff
@@ -0,0 +1,75 @@
+"""A small utility module dedicated to detecting whether or not the `--stamp` flag is enabled
+
+This module can be removed likely after the following PRs ar addressed:
+- https://github.com/bazelbuild/bazel/issues/11164
+"""
+
+load("//rust/private:utils.bzl", "dedent")
+
+StampSettingInfo = provider(
+    doc = "Information about the `--stamp` command line flag",
+    fields = {
+        "value": "bool: Whether or not the `--stamp` flag was enabled",
+    },
+)
+
+def _stamp_build_setting_impl(ctx):
+    return StampSettingInfo(value = ctx.attr.value)
+
+_stamp_build_setting = rule(
+    doc = dedent("""\
+        Whether to encode build information into the binary. Possible values:
+
+        - stamp = 1: Always stamp the build information into the binary, even in [--nostamp][stamp] builds. \
+        This setting should be avoided, since it potentially kills remote caching for the binary and \
+        any downstream actions that depend on it.
+        - stamp = 0: Always replace build information by constant values. This gives good build result caching.
+        - stamp = -1: Embedding of build information is controlled by the [--[no]stamp][stamp] flag.
+
+        Stamped binaries are not rebuilt unless their dependencies change.
+        [stamp]: https://docs.bazel.build/versions/main/user-manual.html#flag--stamp
+    """),
+    implementation = _stamp_build_setting_impl,
+    attrs = {
+        "value": attr.bool(
+            doc = "The default value of the stamp build flag",
+            mandatory = True,
+        ),
+    },
+)
+
+def stamp_build_setting(name, visibility = ["//visibility:public"]):
+    native.config_setting(
```

Comment:
I don't have any issue with this implementation, but here https://docs.bazel.build/versions/main/skylark/lib/transition.html#transition I see that `//command_line_option:cpu` works for transitions for accessing `--cpu`, so maybe `//command_line_option:stamp` works as well and can be used as any other build settings. May be worth trying out.

Reply:
This implementation was inspired by [rules_pkg](https://github.com/bazelbuild/rules_pkg/blob/0.5.1/pkg/private/BUILD#L48-L51) which has been in place for some time. I'd rather stick with the same approach since it seems to have withstood the test of time (so far).

Reply:
Ack. But by trying to use `//command_line_option:stamp` you may discover that it's already possible to access value of `--stamp` flag from Starlark today, negating the need for the workaround.

Reply:
(But if I had to bet, I'd bet on `//command_line_option` thing not working as a normal build setting, I'd expect the label only exists in the context of configuration transitions)

Reply:
These two options may not be what you were thinking but demonstrating that they don't work.

```diff
diff --git a/rust/private/stamp.bzl b/rust/private/stamp.bzl
index 1a7cab6..3292225 100644
--- a/rust/private/stamp.bzl
+++ b/rust/private/stamp.bzl
@@ -48,7 +48,7 @@ def stamp_build_setting(name, visibility = ["//visibility:public"]):
     _stamp_build_setting(
         name = name,
         value = select({
-            ":stamp_detect": True,
+            "//command_line_option:stamp": True,
             "//conditions:default": False,
         }),
         visibility = visibility,
```
```
ERROR: Analysis of target '//rust/private:stamp' failed; build aborted: no such package 'command_line_option': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.
```

```diff
diff --git a/rust/private/stamp.bzl b/rust/private/stamp.bzl
index 1a7cab6..7f0f963 100644
--- a/rust/private/stamp.bzl
+++ b/rust/private/stamp.bzl
@@ -41,7 +41,7 @@ _stamp_build_setting = rule(
 def stamp_build_setting(name, visibility = ["//visibility:public"]):
     native.config_setting(
         name = "stamp_detect",
-        values = {"stamp": "1"},
+        values = {"//command_line_option:stamp": "1"},
         visibility = visibility,
     )
```
```
ERROR: /rules_rust/rust/private/BUILD.bazel:15:20: in values attribute of config_setting rule //rust/private:stamp_detect: error while parsing configuration settings: unknown option: '//command_line_option:stamp'. Since this rule was created by the macro 'stamp_build_setting', the error might have been caused by the macro implementation
```

I've used `//command_line_option` with transitions (exactly what you linked to earlier) but I feel the build setting is simpler than transitions and would opt to keep this as is.

Reply:
This is something I'm happy to make another PR for if I can understand the appropriate change. Since this is an exciting change for some, I'm gonna merge this since ultimately this is an implementation detail that does not impact the API.